### ISSUE TITLE
Regression on: Reentrant scanners #6932

### ISSUE THREAD

**Issue by albert-github** [doxygen/doxygen#6982]
In the pull request #6932 a number of scanners are made reentrant but on a number os systems this leads to the situation that the system now runs out of memory.
When we compare for the mesa package (e.g. from https://fossies.org/linux/misc/mesa-19.0.4.tar.gz or the official site https://www.mesa3d.org/) or the k3b package(https://fossies.org/linux/misc/k3b-19.04.1.tar.gz or the official site https://www.k3b.org/about) on a windows system I get:

mesa:
```
Preprocessing D:/Fossies/mesa-19.0.4/src/gallium/auxiliary/util/u_format_table.c...
out of dynamic memory in constexpYY_create_buffer()

```
and with k3b-19.04.1:
```
Searching for member function documentation...
out of dynamic memory in declinfoYYlex()
```

when running the current master and also the version of May 4, 2019 17:03, when running the version of May 4, 2019 13:51 the both packages don't give these error messages (k3b completes, mesa runs on my system out of memory, though I have from the Fossies administrator the confirmation that the process finishes).

**Comment by jschleus** [comment]
Here as a small addition a plot showing against the processing time in seconds the used percentage memory usage of two different doxygen runs for the package mesa 19.0.4 using doxygen binaries built with the above mentioned master versions. The values were determined via the commend "ps"  on a Linux system with 64 GB memory. 

![image](https://user-images.githubusercontent.com/20492485/57573543-6f390600-7429-11e9-94f0-6b56b4ccea49.png)

Without knowing anything about the internals of the doxygen process one may divide three ranges:

1) Seconds 0 -14: The May 4, 2019 17:03 version takes roughly the factor 4 more memory.
2) Seconds 14-48: The increase is nearly similar for both versions.
3) Seconds 48-58: The May 4, 2019 17:03 version shows a dramatical increase of memory usage (was then killed by the monitoring script since the limit of  50% was exceeded). The May 4, 2019 13:51 version ended without errors after 950 seconds with a memory usage of  4.5%.

**Comment by doxygen** [comment]
Looks like the memory allocated for the scanners, is now done per invocation (instead of once globally) and freeing the memory was not done causing a huge memory leak. I've fixed this with this [commit](https://github.com/doxygen/doxygen/commit/723b842ca1de2f8565717d7099f8ff61b63f6026).

**Comment by albert-github** [comment]
I did a quick test with k3b-19.04.1 (and also with apt-1.7.5 that also crashed) and they completed already. The mesa passes the reported part as well and is now in the dot image generation phase so should complete without problems. mesa takes still a lot of memory but this is due to the size of the program. 
The memory consumption does looks like it is back to normal.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
